awx/main/tasks/system.py

```diff
@@ -245,6 +245,37 @@ def handle_setting_changes(setting_keys):
         reconfigure_rsyslog()
 
 
+@task(queue=get_local_queuename)
+def cancel_control_process(celery_task_id):
+    """Triggers special action in awx.main.dispatch.pool, this is a placeholder"""
+    pass
+
+
+@task(queue=get_local_queuename)
+def cancel_unified_job(unified_job_id):
+    """
+    This method exits to assure cancelation of jobs which had not yet been assigned
+    a celery_task_id, which should be pending and waiting jobs
+    """
+    try:
+        unified_job = UnifiedJob.objects.get(pk=unified_job_id)
+    except UnifiedJob.DoesNotExist:
+        logger.info(f'Job id {unified_job_id} has been deleted, aborting cancel')
+        return
+    while unified_job.status in ACTIVE_STATES:
+        if unified_job.celery_task_id:
+            cancel_control_process.apply_async([unified_job.celery_task_id], queue=unified_job.get_queue_name())
+            logger.warning(f'sigterm issued to {unified_job.log_format} after it obtained a task id')
+            return
+        try:
+            unified_job.refresh_from_db(fields=['status', 'controller_node', 'execution_node'])
+        except unified_job.DoesNotExist:
+            logger.info(f'Job id {unified_job_id} has been deleted, cancel aborted')
+            return
+        time.sleep(1)
+    logger.info(f'{unified_job.log_format} stopped before obtaining a task id, sigterm not needed')
```

Comment:
I'm increasingly thinking this method `cancel_unified_job` shouldn't exist. The race condition it exists to address (which is pending/waiting jobs) shouldn't actually be a problem, as the `cancel_flag` can be relied on as the signaling methods in the cases that matter.

There are much more narrow race conditions to worry about, like when the job is "running" but has not yet been submitted to receptor. Again, this would be the wrong place to handle that. After we set `raise_exception`, we should check if the flag has been set, and treat it as a cancel if it has been. By itself, that should be a complete solution to the potential race conditions.